### ISSUE TITLE
Support of Octavia Availability Zone 

### ISSUE THREAD

**Issue by ITD27M01** [gophercloud/gophercloud#2021]
The Ussuri release of Octavia now supports the availability zone feature where operators can specify some useful capabilities such as the compute availability zone for the amphora instances. For large distributed clouds (enterprise and public), this is a must-have feature that not only improves reliability but also allows operators to use their hardware efficiently. Effectively and according to the tasks. Currently, we have to create a separate zone only for load balancers (with one huge lb-mng network) and all cloud projects use such a hardware pool which sometimes may experience conflicts and resources overflow. This also leads to one point of failure.

I want to suggest to add the "availability-zone" parameters which will be used in Octavia LB request [1]

```
{
    "loadbalancer": {
        "name": "best_load_balancer",
        "provider": "amphora",
        "availability_zone": "my_az",
        "vip_subnet_id": "d4af86e1-0051-488c-b7a0-527f97490c9a",
    }
}
```
The same logic as for flavor-id: [2]

[1] https://docs.openstack.org/api-ref/load-balancer/v2/?expanded=create-a-load-balancer-detail#create-a-load-balancer
[2] https://github.com/gophercloud/gophercloud/blob/master/openstack/loadbalancer/v2/loadbalancers/requests.go#L112

**Comment by jtopjian** [comment]
@ITD27M01 Thanks for reporting this. If it's a valid feature in OpenStack/Octavia, then it can definitely be added here. If you'd like to submit a PR for this, please take a look at our [contributor tutorial](https://github.com/gophercloud/gophercloud/tree/master/docs/contributor-tutorial) for how to get started.

**Comment by ITD27M01** [comment]
Hello @jtopjian and @lingxiankong

Is there any policy to support the backward compatibility of previous versions of Octavia service? I ask it because of @jichenjc question in https://github.com/kubernetes/cloud-provider-openstack/issues/1223

For example, [Octavia Tags](https://github.com/gophercloud/gophercloud/pull/1406) are added in [2.5](https://docs.openstack.org/api-ref/load-balancer/v2/#filtering-by-tags)

Should I worry about this and add code to check the API version? In the case of absence of ability there is just 404 response:

```
GET call to https://octavia-public-testos.example.com:9876/v2.0/lbaas/availabilityzones used request id req-cfd21aff-99e2-4404-83b9-a7230020c891
Request returned failure status: 404
Unrecognized schema in response body. (HTTP 404) (Request-ID: req-cfd21aff-99e2-4404-83b9-a7230020c891)
Traceback (most recent call last):
  File "/Users/igor.tiunov/anaconda3/envs/openstack/lib/python3.8/site-packages/cliff/app.py", line 400, in run_subcommand
    result = cmd.run(parsed_args)
  File "/Users/igor.tiunov/anaconda3/envs/openstack/lib/python3.8/site-packages/cliff/display.py", line 117, in run
    column_names, data = self.take_action(parsed_args)
  File "/Users/igor.tiunov/anaconda3/envs/openstack/lib/python3.8/site-packages/octaviaclient/osc/v2/availabilityzone.py", line 141, in take_action
    data = self.app.client_manager.load_balancer.availabilityzone_list(
  File "/Users/igor.tiunov/anaconda3/envs/openstack/lib/python3.8/site-packages/octaviaclient/api/v2/octavia.py", line 982, in availabilityzone_list
    response = self._list(path=url, **kwargs)
  File "/Users/igor.tiunov/anaconda3/envs/openstack/lib/python3.8/site-packages/osc_lib/api/api.py", line 227, in list
    ret = self._request(
  File "/Users/igor.tiunov/anaconda3/envs/openstack/lib/python3.8/site-packages/osc_lib/api/api.py", line 139, in _request
    return session.request(url, method, **kwargs)
  File "/Users/igor.tiunov/anaconda3/envs/openstack/lib/python3.8/site-packages/keystoneauth1/session.py", line 976, in request
    raise exceptions.from_response(resp, method, url)
keystoneauth1.exceptions.http.NotFound: Unrecognized schema in response body. (HTTP 404) (Request-ID: req-cfd21aff-99e2-4404-83b9-a7230020c891)
clean_up ListAvailabilityzone: Unrecognized schema in response body. (HTTP 404) (Request-ID: req-cfd21aff-99e2-4404-83b9-a7230020c891)
END return value: 1
```

**Comment by jtopjian** [comment]
@ITD27M01 If 2.5 should be considered a microversion within the Octavia API, then [this document](https://github.com/gophercloud/gophercloud/blob/master/docs/MICROVERSIONS.md) should help with how to add support for tags.

**Comment by ITD27M01** [comment]
@jtopjian So, if Octavia doesn't support microversions and doesn't react on any http headrs (like X-OpenStack-Nova-API-Version for Nova Compute service), should I implement such checks somewhere within the gophercloud code?

**Comment by jtopjian** [comment]
If it is an "additive" feature to the API major version, then the best way of implementing it was to make sure that the fields were omitted entirely from the JSON request if they were not specified.

It is then up to the application that uses Gophercloud check the API version of the Octavia service being offered by the OpenStack cloud. Gophercloud should implement the [means](https://github.com/gophercloud/gophercloud/tree/master/openstack/loadbalancer/v2/apiversions) to do these checks, but not do the actual check itself.

**Comment by jichenjc** [comment]
I think the microversion is in server side, the client side can indicate which version to use explicitly ..
for example

 used 2.60 version API here
https://github.com/kubernetes/cloud-provider-openstack/blob/master/pkg/csi/cinder/openstack/openstack_volumes.go#L170

then can switch back to default version ,of course, only use microversion when we are sure it must be used

**Comment by jtopjian** [comment]
> I think the microversion is in server side, the client side can indicate which version to use explicitly ..
for example

Right, however the example you gave is for setting the microversion in the Compute/Nova service. If I understand @ITD27M01 correctly, Octavia doesn't work this way.

Here are the current microversion-enabled services that Gophercloud knows about: https://github.com/gophercloud/gophercloud/blob/41c3197ccfafa597c7a059949edb152a6b9910a2/service_client.go#L124-L141

Perhaps Octavia uses the generic `OpenStack-API-Version` header? But this is beyond my knowledge of Octavia.

**Comment by ITD27M01** [comment]
Hi

I've found that AZ support added in v2.14: https://github.com/openstack/octavia/blob/master/octavia/api/root_controller.py#L109

But I didn't find the microversions support [1] in the way Nova, Cinder, and other projects do. It is just v2 and all subsequent minor versions are part of it [2]. I assume they haven't stepped on the rake yet or too lazy to implement it: https://storyboard.openstack.org/#!/story/1596655

I agree with @jtopjian that the best way of implementing it is to make sure that the fields are omitted entirely from the JSON request if they were not specified. Besides, for the application, there is the possibility of checking the version if [it needs to be](https://github.com/gophercloud/gophercloud/tree/master/openstack/loadbalancer/v2/apiversions).

@jichenjc Just and reminder how microversion is used:

The client can specify the minor version it wants to use:
```
openstack server migrate --live-migration --host srv-os-kvm01 --os-compute-api-version 2.99 63a5d8da-1f8d-495b-9bc2-b8bbb019c829 --debug
```
The first check will occur on client side:
```
  File "/Users/igor.tiunov/anaconda3/envs/openstack/lib/python3.8/site-packages/openstackclient/compute/client.py", line 152, in check_api_version
    raise exceptions.CommandError(msg)
osc_lib.exceptions.CommandError: versions supported by client: 2.1 - 2.87
```

And the upper version is checked on server side:
```
openstack server migrate --live-migration --host srv-os-kvm01 --os-compute-api-version 2.87 63a5d8da-1f8d-495b-9bc2-b8bbb019c829 --debug

RESP BODY: {"computeFault": {"code": 406, "message": "Version 2.87 is not supported by the API. Minimum is 2.1 and maximum is 2.79."}}
```

So, when you specify the 2.60 version API here
https://github.com/kubernetes/cloud-provider-openstack/blob/master/pkg/csi/cinder/openstack/openstack_volumes.go#L170

you dont add the backward compatibility, there is just convenient way to handling the corresponding error. The backward compatibility is managed by service (Nova/Cinder, etc...) server-side - they specify default minor version and require the client to raise this default version if it requires: https://github.com/openstack/nova/blob/master/nova/api/openstack/api_version_request.py#L248

[1] https://github.com/openstack/octavia/commit/8ae6bc3697434aa2c8676afbff3d4938676a35a5
[2] https://github.com/openstack/octavia/blob/master/octavia/api/root_controller.py#L70

**Comment by jtopjian** [comment]
Implemented in #2026
